Homepage

AsianHam opened the pull request:
basic homepage look

AsianHam committed:
basic fetch

Added a function to fetch all recipes

AsianHam committed:
Basic Home Page Setup

UI now properly pulls data from API

Setup base template to generate antd collapse content from the API.

AsianHam committed:
Images

added basic image support onto the home page

AsianHam committed:
dynamic image attempt

thumbnail properly disappears when expanding panel

thumbnail does not reappear when collapsing panel

AsianHam committed:
expanded recipe card template

created base template for a recipe card on the homepage once expanded

AsianHam committed:
Images fixed

Thumbnail image properly hides and shows depending on the expansion of the recipe card

Updated the Recipe interface for better state tracking

Recommendations properly show a picture of the recipe now

AsianHam committed:
ingredient update

updated ingredients array to include a "starred" option

updated the ingredients box to only show starred ingredients

AsianHam committed:
code cleanup

removed commented lines and functions

jackson-miller-tbol commented on src/App.tsx:
```diff
@@ -1,24 +1,154 @@
+  const [page, setPage] = useState<Recipe[]>();
+
+  const getAllRecipes = async () => {
+    const response = await fetch("http://localhost:3001/recipe");
```
Make the base URL here an environment variable.  We'll need to be able to set that in different environments

jackson-miller-tbol commented on src/App.tsx:
```diff
@@ -1,24 +1,154 @@
+    const data: Recipe[] = await response.json();
+    if (data !== undefined) {
+      data.map((recipe) => {
+        if (recipe.thumbnail.image === "") {
```
Will this always be an empty string if it's not populated?  Or will it be undefined?

AsianHam replied:
the way the db is currently setup and the way the edit page will be setup will ensure that image is always a string

jackson-miller-tbol commented on src/App.tsx:
```diff
@@ -1,24 +1,154 @@
+import { DeleteOutlined, EditOutlined } from "@ant-design/icons";
 
 function App() {
+  const [page, setPage] = useState<Recipe[]>();
```
I feel like you should rename this

jackson-miller-tbol reviewed (changes requested):
I'll look more closelier when I get home but I'd see if there are ways you can avoid null checking "page" everywhere.  It'll make the code cleaner

AsianHam committed:
pr changes

removed borders around divs

updated api fetch url to an environment variable

added stylized borders in recipe card for additional readability

removed repeated undefined checks

renamed state variable

Jacksondr5 commented on package.json:
```diff
@@ -3,13 +3,17 @@
     "@testing-library/jest-dom": "^5.14.1",
     "@testing-library/react": "^13.0.0",
     "@testing-library/user-event": "^13.2.1",
     "@types/jest": "^27.0.1",
```
These @types can move to the devDependencies object

Jacksondr5 commented on src/App.tsx:
```diff
@@ -1,24 +1,157 @@
+import { DeleteOutlined, EditOutlined } from "@ant-design/icons";
 
 function App() {
+  const url = `${process.env.REACT_APP_URL}`;
```
I would rename the env var to API_URL or something, since it's not a React app

Jacksondr5 replied:
This also assumes that the env var has the route in it (/recipe), which is generally not how people do things.  While you may only end up with one route, if you had more than one this wouldn't work well

AsianHam replied:
The variable has to be named REACT_APP_XXX, the variable doesn't get picked up without it.

https://create-react-app.dev/docs/adding-custom-environment-variables/

![Screenshot 2022-07-06 092950](https://user-images.githubusercontent.com/25393502/177561696-a8517eb4-7520-478c-b2ff-64739d4af1d0.png)

jackson-miller-tbol replied:
Ah OK, in that case can you call it `REACT_APP_MONGO_URL`?

AsianHam replied:
I did change it to REACT_APP_API_URL,  but yeah i can change it to mongo_url if you like that better

Jacksondr5 replied:
oh duh its an api url, ok that works

Jacksondr5 commented on src/App.tsx:
```diff
@@ -1,24 +1,157 @@
 function App() {
+  const url = `${process.env.REACT_APP_URL}`;
+  if (url === undefined) {
+    throw new Error("invalid URL");
```
I would make this error message more helpful.  You'll only really see this error in 2 scenarios:
1. A new dev pulls this down and didn't set their local env var
2. You're deploying this and forgot to set the env var
In both cases if you don't intimately know the code, the error message doesn't tell you what's really wrong and you have to look through the code to find the problem.  If you said that the environment variable was missing, then the person troubleshooting the issue immediately knows what to do

Jacksondr5 commented on src/App.tsx:
```diff
@@ -1,24 +1,157 @@
 
 function App() {
+  const url = `${process.env.REACT_APP_URL}`;
+  if (url === undefined) {
```
```suggestion
  if (!url) {
```
This is generally the syntax for what you're doing here.  It's technically more inclusive

Jacksondr5 commented on src/App.tsx:
```diff
@@ -1,24 +1,157 @@
+          <div className="Recipes" key={index}>
+            <>
+              {item.thumbnail.show && (
+                <Image width={150} src={item.thumbnail.image} preview={false} />
```
I think I actually kinda like the preview ability here.  I don't think it's desirable on the recommendations though, since you'd want to tap to go to the recipe on those

Jacksondr5 commented on src/App.tsx:
```diff
@@ -1,24 +1,157 @@
-          Learn React
-        </a>
-      </header>
+      <h1>Recipe App</h1>
```
I feel like you can just remove this.  We all know what we're looking at here

Jacksondr5 commented on src/App.tsx:
```diff
@@ -1,24 +1,157 @@
+    }
+  };
+
+  const displayIngredient = (ingredient: string, starred: boolean) => {
```
This function is redundant.  You can just do `ingredient.starred` where you call this function

AsianHam replied:
So don't call the function at all? Am I doing conditional rendering like with the thumbnail? i.e. {ingredient.starred && ingredient.ingredient}

Jacksondr5 commented on src/recipe.ts:
```diff
@@ -0,0 +1,31 @@
+  name: string;
+  thumbnail: Thumbnail;
+  description: string;
+  link: string[];
```
I realize this is because of the schema changes, but you should just make this an `number[]` so that you don't have to do all those parseInt calls in app.tsx

Jacksondr5 commented on src/App.tsx:
```diff
@@ -1,24 +1,157 @@
+  const onChange = (
+    e: string | string[],
+    index: number,
+    image: string,
```
I think `image` and `show` are redundant here since you just get the recipe on line 41 anyway

AsianHam replied:
No, they're the image and show of the current recipe. I can remove them but I'll need to replace the code with an id search.

![image](https://user-images.githubusercontent.com/25393502/177616704-1cb7e435-6ac5-410a-8ccc-bf692eaec038.png)

Jacksondr5 commented on src/App.tsx:
```diff
@@ -1,24 +1,157 @@
+            <div className="RecipeBox">
+              <Collapse
+                onChange={(e) =>
+                  onChange(e, index, item.thumbnail.image, item.thumbnail.show)
```
I don't think you need to pass the event in if you aren't using it.  You can just pass `() => onChange(index)`

Jacksondr5 commented on src/App.css:
```diff
@@ -1,38 +1,106 @@
-.App-logo {
-  height: 40vmin;
-  pointer-events: none;
+.Ipad {
```
I believe css class names are supposed to be `hyphen-case`

AsianHam replied:
Do you want me to change all the css class names?

Jacksondr5 replied:
yes

Jacksondr5 commented on src/App.tsx:
```diff
@@ -1,24 +1,157 @@
+                  <div className="FirstHalf">
+                    <div className="IngredientBox">
+                      <h2>Ingredients</h2>
+                      {item.ingredients.map((ingred, ingredIndex) => (
```
Please change this to `ingredient`.  You generally don't want to abbreviate variable names except in certain rare circumstances

Jacksondr5 commented on src/App.tsx:
```diff
@@ -1,24 +1,157 @@
+      <h1>Recipe App</h1>
+
+      <div className="Ipad">
+        {allRecipes.map((item, index) => (
```
I would change `item` to `recipe` here

Jacksondr5 commented on src/App.tsx:
```diff
@@ -1,24 +1,157 @@
+                    <h2>Recommendations</h2>
+                  </div>
+                  <div className="RecommendationsBox">
+                    {item.link.map((rec, recIndex) => (
```
I would change `rec` to `linkedRecipeId`

Jacksondr5 commented on src/App.tsx:
```diff
@@ -1,24 +1,157 @@
+    const response = await fetch(url);
+    const data: Recipe[] = await response.json();
+    if (data === undefined) {
+      throw new Error("bad api call");
```
Again, I'd make a more helpful error message here

Jacksondr5 commented on src/App.tsx:
```diff
@@ -1,24 +1,157 @@
+    const data: Recipe[] = await response.json();
+    if (data === undefined) {
+      throw new Error("bad api call");
+    } else {
```
I don't think the else is necessary here, since there isn't code executed after it

Jacksondr5 reviewed (changes requested):
While there are a lot of changes requested here, they're nearly all just more efficient ways of doing things or conventions, not major issues with the code.  Really good work overall

Jacksondr5 commented:
![image](https://user-images.githubusercontent.com/9014034/177459404-0645b0f4-465b-4fcc-848c-dac0e74c3aca.png)
The headings are different for ingredients and recommendations.  I'd recommend styling them the same.  You could experiment with moving the metadata up to the description box (long descriptions will get pushed down, but more room for ingredients) and centering the ingredients header.  Or you could try changing the alignment and spacing on the recommendations section.  In general, you have a mix of centered single column, centered double column, and left aligned.  You should decide on one of those 3.

Also, the recipe title should have some kind of styling to it.  Either bolding or larger font or something.  It shouldn't be the same as the ingredient list.

AsianHam committed:
PR changes

add .env.example

changed url env to include the route

changed null checks to be more inclusive

added more descriptive error messages

renamed certain variables

shifted metadata box next to description box

bolded and increased recipe name in header

Jacksondr5 commented on src/App.tsx:
```diff
@@ -1,24 +1,179 @@
+import { DeleteOutlined, EditOutlined } from "@ant-design/icons";
 
 function App() {
+  const baseUrl = `${process.env.REACT_APP_API_URL}`;
```
This actually swallows the error because if the env var is undefined, you the string "undefined"

AsianHam committed:
more pr updates

renamed css classes

updated env to catch empty env

removed unnecessary function params

removed title bar

removed redundant functions

AsianHam committed:
interface update

changed an interface export name to singular

AsianHam committed:
interface update

Removed a property from the recipe interface that doesn't get updated in the database

intersected a new interface to house the show property

Jacksondr5 commented on src/App.tsx:
```diff
@@ -6,16 +6,23 @@ import { Button, Collapse, Image } from "antd";
 import { DeleteOutlined, EditOutlined } from "@ant-design/icons";
 
 function App() {
+  type uiRecipe = Recipe & {
```
Should be `UiRecipe`
